cpp/include/raft/neighbors/detail/ivf_pq_build.cuh

```diff
@@ -822,140 +757,178 @@ void process_and_fill_codes(raft::device_resources const& handle,
-                       stream);
-  switch (source.codebook_kind()) {
-    case codebook_gen::PER_SUBSPACE: {
-      copy(target.pq_centers().data_handle(),
-           source.pq_centers().data_handle(),
-           source.pq_centers().size(),
-           stream);
-    } break;
-    case codebook_gen::PER_CLUSTER: {
-      raft::matrix::gather(source.pq_centers().data_handle(),
-                           static_cast<IdxT>(source.pq_book_size() * source.pq_len()),
-                           static_cast<IdxT>(source.n_lists()),
-                           cluster_ordering,
-                           static_cast<IdxT>(target.n_lists()),
-                           target.pq_centers().data_handle(),
-                           stream);
-    } break;
-    default: RAFT_FAIL("Unreachable code");
-  }
+  // Copy shared pointers
+  target.lists() = source.lists();
 
-  // Fill the data with the old clusters.
-  if (source.size() > 0) {
-    std::vector<IdxT> target_cluster_offsets(target.n_lists() + 1);
-    std::vector<IdxT> source_cluster_offsets(target.n_lists() + 1);
-    std::vector<uint32_t> source_cluster_sizes(target.n_lists());
-    copy(target_cluster_offsets.data(),
-         target.list_offsets().data_handle(),
-         target.list_offsets().size(),
-         stream);
-    copy(source_cluster_offsets.data(),
-         source.list_offsets().data_handle(),
-         source.list_offsets().size(),
-         stream);
-    copy(source_cluster_sizes.data(),
-         source.list_sizes().data_handle(),
-         source.list_sizes().size(),
-         stream);
-    stream.synchronize();
-    auto data_exts = target.pq_dataset().extents();
-    auto data_unit = size_t(data_exts.extent(3)) * size_t(data_exts.extent(1));
-    auto data_mod  = size_t(data_exts.extent(2));
-    for (uint32_t l = 0; l < target.n_lists(); l++) {
-      auto k                   = cluster_ordering[l];
-      auto source_cluster_size = source_cluster_sizes[k];
-      if (source_cluster_size > 0) {
-        copy(target.indices().data_handle() + target_cluster_offsets[l],
-             source.indices().data_handle() + source_cluster_offsets[k],
-             source_cluster_size,
-             stream);
-        copy(target.pq_dataset().data_handle() + target_cluster_offsets[l] * data_unit,
-             source.pq_dataset().data_handle() + source_cluster_offsets[k] * data_unit,
-             round_up_safe<size_t>(source_cluster_size, data_mod) * data_unit,
-             stream);
-      }
-    }
-  }
+  // Make sure the device pointers point to the new lists
+  recompute_internal_state(res, target);
+
+  return target;
 }
 
-/** See raft::spatial::knn::ivf_pq::extend docs */
+/**
+ * Extend the index in-place.
+ * See raft::spatial::knn::ivf_pq::extend docs.
+ */
 template <typename T, typename IdxT>
-auto extend(raft::device_resources const& handle,
-            const index<IdxT>& orig_index,
+void extend(raft::device_resources const& handle,
```

Comment:
For the faiss integration, we need to support both extending an index with a new set of vectors as well as being able to write vectors directly into the lists (for the case of converting to/from a faiss index on the cpu/gpu.

It would help if we had a utility function to encode and write the (non-interleaved) flat indices from a block of memory directly to a specific list.

Reply:
Makes sense! I'd suggest we add this in a follow-up PR though. The current implementation is optimized to go over all input data in a single pass, so there's no obvious separation between the sorting input in the clusters and adding them to lists.

Reply:
Yep I'm definitely okay doing that in a follow up.